### DISCUSSION TITLE
schema of union type is `anyOf` ?

### DISCUSSION THREAD

**Discussion by Yasu-umi** [pydantic/pydantic#4744] [Question] [answered]
I tried to get schema of union type as below.
I expect `oneOf` but got `anyOf`.
[this improvement](https://github.com/pydantic/pydantic/pull/4335) in 1.10 release resolved this `anyOf` issue, I think.
Is this expected behavior? or should I open new issue for bug fix ?

```python
from pydantic import BaseModel
class A(BaseModel):
    a: int

class B(BaseModel):
    b: str

class D(BaseModel):
    __root__: A | B

print(D.schema())
```
output
```python
{'title': 'D', 'anyOf': [{'$ref': '#/definitions/A'}, {'$ref': '#/definitions/B'}], 'definitions': {'A': {'title': 'A', 'type': 'object', 'properties': {'a': {'title': 'A', 'type': 'integer'}}, 'required': ['a']}, 'B': {'title': 'B', 'type': 'object', 'properties': {'b': {'title': 'B', 'type': 'string'}}, 'required': ['b']}}}
```
pydantic version: 1.10.2

**Comment by rafalkrupinski** [comment, marked as answer]
That's quite a typical mistake.
`oneOf` means that the object validates against only one of the listed schemas, and it fails validation against all others. That's outside of scope of Union.

Consider this example:
```python
class A(BaseModel):
    a: str

class B(BaseModel):
    a: str = Field(max_length=3)

pydantic.parse_obj_as(Union[A, B], d)
```

If d is `dict(a='1234)` then it's a valid A but not B, so it validates against `oneOf: [{$ref:"A"}, {$ref: "B"}]` (as well as analogous `anyOf`).
But `dict(a='123')` validates against both A and B, which is OK for `Union[A, B]` but wouldn't validate against `oneOf`, only `anyOf`.

Hope I've explained it clearly enough :)

**Reply by rafalkrupinski** [reply]
And BTW `anyOf` isn't 1-to-1 representation of `Union`, as `anyOf` allows the object to validate against more than one of sub-schemas, which isn't the case with `Union`. Even if your model allows extra attributes, there's no (builtin) way to validate them.

**Reply by Yasu-umi** [reply]
Thanks for details explanation!
now I understand oneof is subset of anyof.
